### ISSUE TITLE
Signal handler doesn't run if signal received at the wrong time

### ISSUE THREAD

**Issue by talex5** [ocaml/ocaml#13189]
When a signal arrives, the C function `handle_signal` is called, which calls `caml_record_signal`. This just sets a flag indicating that the corresponding OCaml handler should run at the next opportunity.

- If OCaml code is running, it will notice the flag after a bit and run the handler.
- If a system call is in progress, it will return EINTR. The C stub should call `caml_leave_blocking_section` and the handler will run.
- However, if a C stub has called `caml_enter_blocking_section` but not yet started the system call, then the system call will start and can block for any amount of time.

Test case:
```ocaml
external test : unit -> unit = "test_c_stub"

(* If the alarm goes off after [test] has entered its blocking section and
   before it calls [nanosleep], this won't run. *)
let handle _ =
  print_endline "Got signal";
  exit 0

let () =
  Sys.(set_signal sigalrm) (Signal_handle handle);
  print_endline "Scheduling SIGALRM 1ms from now";
  ignore (Unix.setitimer ITIMER_REAL { it_interval = 0.; it_value = 0.001 } : Unix.interval_timer_status);
  test ()
```

Here's a C stub with a delay to make the problem easier to see (I compiled with `-O0` so gcc doesn't optimise the loop out):
```c
#include <caml/mlvalues.h>
#include <caml/signals.h>
#include "caml/unixsupport.h"

#include <errno.h>
#include <time.h>

CAMLprim value test_c_stub(value v_unit)
{
  struct timespec t;
  int ret;
  t.tv_sec = (time_t) 10000;
  t.tv_nsec = 0;
  caml_enter_blocking_section();
  printf("Minor delay to make problem easier to see...\n");
  for (long i = 0; i < 10000000; i++) { }
  printf("Calling nanosleep...\n");
  ret = nanosleep(&t, &t);
  caml_leave_blocking_section();
  if (ret == -1) caml_uerror("test_c_stub", Nothing);
  return Val_unit;
}
```

The output I get is:
```
Scheduling SIGALRM 1ms from now   
Minor delay to make problem easier to see...
Calling nanosleep...
[ hangs ]
```

I haven't seen this happen in the wild, but I suspect a fix for this would help me in some other places. I guess the solution would be to have the C signal handler write to an eventfd, and have anything running an event loop select on that (e.g. the above `nanosleep` would become a `select` with a timeout).

My actual problem is that `_io_uring_get_cqe` loops if it gets interrupted after submitting some requests but before getting any responses. If OCaml wrote to an eventfd then I could get Eio to watch that and it wouldn't be a problem.

I can mostly work around the problem by making my own C signal handler, but it would be good to work correctly with code using `Sys.set_signal` too.

**Comment by gadmm** [comment]
The race between signals and blocking syscalls has been mentioned before on this repository, with a suggestion to look into the self-pipe trick (I think, a similar way to do what you suggest but more portable). Thanks for the issue because this helps keep track of it.

Note that `caml_enter_blocking_section` checks for signals right before executing the blocking section, in order to reduce the likelihood of the bad race situation, without eliminating it.

Now the concrete problem you face as described in the issue linked above is that io_uring does their own restart in case of EINTR. As you describe, one solution would be to fix io_uring so that it propagates EINTR.

Can you please describe a bit more the other solution you have in mind, that would let you fix the issue on your side without changing io_uring?

It is not clear to me that the two problems are related as strongly as you wrote, but a solution for the second one that involves improving the OCaml interface for signal handlers with a fd for signals could be a step to a solution for the first one.

---

Some additional details: I think the fd should not record pending signals, since we don't know which thread is ready to handle it. It should record whether a signal just arrived and the thread has not yet looked at the pending OCaml signal handlers, corresponding to the situation `young_limit==UINTNAT_MAX` as set from the C signal handler. It is the exact same condition as when deciding to check again for signals inside `caml_enter_blocking_section`. I also fear that there would need to be one such fd per domain...

**Comment by talex5** [comment]
Maybe what I really want is a way to register a C function to be called by `caml_record_signal`. IO libraries like Eio generally have their own ways to wake up the event loop and can just call that.

> Can you please describe a bit more the other solution you have in mind, that would let you fix the issue on your side without changing io_uring?

The way it fixes C libraries restarting on EINTR is:

1. We register an extra signal callback (possibly just for the duration of the C stub).
2. The C stub enters the blocking section.
3. The C stub calls the C library (uring, libuv, glib, etc), which calls `select` or similar.
4. A signal occurs, which calls the registered callback.
5. The callback writes to the eventfd / wake-up pipe.
6. `select` returns `EINTR` and the C library retries it.
7. The second `select` syscall returns immediately saying that the eventfd is ready for reading.

So this way it doesn't matter if a C library retries on EINTR.

**Comment by talex5** [comment]
> I also fear that there would need to be one such fd per domain

I don't see why. If fact, we possibly only need one domain to be using this system at all:
- If any domain is running OCaml code then the signal handler will run anyway.
- If all domains are blocking, it's fine to always use the same one to handle signals.

**Comment by gadmm** [comment]
The OCaml signal handler will only execute on target thread(s) as defined via `Thread.sigmask`. This is why it is not enough to run the OCaml signal handler inside an arbitrary domain. (This ensures for instance that `Sys.Break` gets raised in the thread you expect.) In addition, you want the target thread (or one of the target thread*s*) to have an ongoing syscall interrupted, so you need a way to ensure that the event is not consumed in parallel by a systhread that is not going to run the signal handler.

But this is for the general problem. Are things different for eio?

**Comment by talex5** [comment]
Good point about signal masks. For most signals (e.g. SIGCHLD) the OCaml handler just triggers an Eio condition and any domains waiting for it will wake up and handle the signal from their event loop when ready. But it's typically different for SIGINT, indeed, since that doesn't wait for the current fiber to yield before being handled so it matters which thread runs the OCaml handler.
